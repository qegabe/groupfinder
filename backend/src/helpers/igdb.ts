import axios from "axios";

const IGDB_URL = "https://api.igdb.com/v4";

/**
 * Checks if the twitch token is valid
 * @returns {Promise<boolean>}
 */
async function validateToken(): Promise<boolean> {
  if (process.env.TWITCH_TOKEN !== undefined) {
    try {
      const res = await axios({
        method: "GET",
        url: `https://id.twitch.tv/oauth2/validate`,
        headers: {
          Authorization: `Bearer ${process.env.TWITCH_TOKEN}`,
        },
      });

      console.log(`Twitch validated with ${res.status}`);
      return true;
    } catch (error) {
      console.error(error);
      return false;
    }
  }

  return false;
}

/**
 * Send a request to IGDB
 * @param endpoint endpoint for request
 * @param data
 * @returns {Promise<any[]>}
 */
async function requestIGDB(endpoint: string, data: string): Promise<any[]> {
  console.log("Sending request to IGDB...");
  try {
    const resp = await axios({
      url: `${IGDB_URL}/${endpoint}`,
      method: "POST",
      headers: {
        Authorization: `Bearer ${process.env.TWITCH_TOKEN}`,
        "Client-ID": process.env.TWITCH_CLIENT_ID,
      },
      data,
    });
    return resp.data;
  } catch (error) {
    //console.error(error);
    console.error(`IGDB Request failed with: ${error.code}`);
  }
}

export { validateToken, requestIGDB };
