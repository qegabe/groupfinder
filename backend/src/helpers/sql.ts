import { BadRequestError } from "./expressError";
import { Filter } from "../types";

const jsToSql = {
  avatarUrl: "avatar_url",
  triviaScore: "trivia_score",
  isPrivate: "is_private",
  startTime: "start_time",
  endTime: "end_time",
  maxMembers: "max_members",
};

/**
 * Creates sql for updating a row
 * @param dataToUpdate
 * @returns {object} { setCols, values }
 */
function sqlForPartialUpdate(dataToUpdate: object) {
  const keys = Object.keys(dataToUpdate);
  if (keys.length === 0) throw new BadRequestError("No data");

  // {username: 'mike', avatarUrl: 'img.jpg'} => ['"username"=$1', '"avatar_url"=$2']
  const cols = keys.map(
    (colName, idx) =>
      `"${jsToSql[colName as keyof object] || colName}"=$${idx + 1}`
  );

  return {
    setCols: cols.join(", "),
    values: Object.values(dataToUpdate),
  };
}

/**
 * Creates sql for filtering a SELECT query
 * @param filter
 * @returns {object} { matchers, values }
 */
function sqlForFiltering(filter: Filter) {
  const keys = Object.keys(filter);

  const matchers = [];
  const values = [];
  for (let key of keys) {
    const i = values.length + 1;
    switch (key) {
      case "title":
        matchers.push(`title ILIKE $${i}`);
        values.push(`%${filter.title}%`);
        break;
      case "startTimeAfter":
        matchers.push(`start_time >= $${i}`);
        values.push(filter.startTimeAfter);
        break;
      case "startTimeBefore":
        matchers.push(`start_time <= $${i}`);
        values.push(filter.startTimeBefore);
        break;
      case "maxSize":
        const m = `id IN (SELECT id
                FROM groups
                LEFT JOIN groupsusers ON groupsusers.group_id = groups.id
                GROUP BY id
                HAVING COUNT(*) <= $${i}
              )`;
        matchers.push(m);
        values.push(filter.maxSize);
        break;
      default:
        break;
    }
  }

  return { matchers, values };
}

/**
 * Creates sql for the columns and values of an INSERT query
 * @param data
 * @returns {object} { colString, valString, values}
 */
function sqlForInserting(data: object) {
  const keys = Object.keys(data);

  const cols: string[] = [];
  const valueIdxs: string[] = [];

  for (let i = 0; i < keys.length; i++) {
    cols.push(jsToSql[keys[i] as keyof object] || keys[i]);
    valueIdxs.push(`$${i + 1}`);
  }

  return {
    colString: `(${cols.join(", ")})`,
    valString: `(${valueIdxs.join(", ")})`,
    values: Object.values(data),
  };
}

export { sqlForPartialUpdate, sqlForFiltering, sqlForInserting };
