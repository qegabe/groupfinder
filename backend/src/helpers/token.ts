import jwt from "jsonwebtoken";
import { SECRET_KEY } from "../config";
import { IUser } from "../types";

function createToken(user: IUser) {
  const payload = {
    username: user.username,
  };

  return jwt.sign(payload, SECRET_KEY);
}

export default createToken;
