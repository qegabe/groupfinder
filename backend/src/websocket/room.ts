import WSUser from "./wsUser";

// in-memory storage of roomNames -> room

const ROOMS = new Map();

/** Room is a collection of listening members; this becomes a "chat room"
 *   where individual users can join/leave/broadcast to.
 */
class Room {
  id: number;
  members: Set<WSUser>;

  static get(roomId: number) {
    if (!ROOMS.has(roomId)) {
      ROOMS.set(roomId, new Room(roomId));
    }

    return ROOMS.get(roomId);
  }

  /** make a new room, starting with empty set of listeners */
  constructor(roomId: number) {
    this.id = roomId;
    this.members = new Set();
  }

  /** member joining a room. */
  join(member: WSUser) {
    this.members.add(member);
  }

  /** member leaving a room. */
  leave(member: WSUser) {
    this.members.delete(member);
  }

  /** send message to all members in a room. */
  broadcast(data: any) {
    for (let member of this.members) {
      if (member.verified) {
        member.send(JSON.stringify(data));
      }
    }
  }
}

export default Room;
