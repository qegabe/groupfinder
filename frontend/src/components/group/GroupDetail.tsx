import React, { useEffect, useState } from "react";
import { Box, Grid, Button, Typography } from "@mui/material";
import GroupFinderApi from "../../api";
import { Link, useParams } from "react-router-dom";
import LoadingSpinner from "../common/LoadingSpinner";
import dayjs from "dayjs";
import duration from "dayjs/plugin/duration";
import relativeTime from "dayjs/plugin/relativeTime";
import { useAppSelector } from "../../hooks/redux";
import GameList from "../game/GameList";
import UserList from "../user/UserList";
dayjs.extend(duration);
dayjs.extend(relativeTime);

function GroupDetail() {
  const { id } = useParams();
  const [group, setGroup] = useState<Group>();
  const user = useAppSelector((s) => s.auth.user);

  useEffect(() => {
    async function loadGroup() {
      setGroup(await GroupFinderApi.getGroup(+(id as string)));
    }
    loadGroup();
  }, [id]);

  if (!group) return <LoadingSpinner />;

  let edit = null;
  if (group.members) {
    if (group.members[user?.username as keyof object]) {
      edit = (
        <Button variant="contained" component={Link} to={`/groups/${id}/edit`}>
          Edit
        </Button>
      );
    }
  }

  return (
    <Box sx={{ display: "grid", justifyItems: "center" }}>
      <Typography variant="h3" sx={{ my: 2, textTransform: "capitalize" }}>
        {group.title}
      </Typography>
      <Box sx={{ my: 2 }}>
        <Typography sx={{ fontStyle: "italic" }}>
          {group.description}
        </Typography>
      </Box>

      <Typography>Starting at: {group.startTime.format("LLL")}</Typography>
      <Typography>
        Lasts {dayjs.duration(group.endTime.diff(group.startTime)).humanize()}
      </Typography>
      <Grid container spacing={2}>
        <Grid item sx={{ display: "grid", justifyContent: "center" }} xs={6}>
          <Box sx={{ my: 2 }}>
            <Typography>Users</Typography>
            <UserList userData={group.members} />
          </Box>
        </Grid>
        <Grid item sx={{ display: "grid", justifyContent: "center" }} xs={6}>
          <Box sx={{ my: 2 }}>
            <Typography>Games</Typography>
            <GameList gameData={group.games} />
          </Box>
        </Grid>
      </Grid>

      {edit}
    </Box>
  );
}

export default GroupDetail;
