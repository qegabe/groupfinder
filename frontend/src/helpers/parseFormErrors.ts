function simplifyError(error: string) {
  if (error.includes("User already exists with username")) {
    return { username: { error: true, helperText: "Username already taken" } };
  }
  return { message: error };
}

function parseFormErrors(error: string) {
  let errArray;
  try {
    errArray = JSON.parse(error);
  } catch {
    return simplifyError(error);
  }

  const errors: any = {};
  if (Array.isArray(errArray)) {
    for (let e of errArray) {
      const prop = e.match(/^instance\.(\w+)/)[1];
      if (prop) {
        if (e.includes("does not meet minimum length")) {
          errors[prop as keyof object] = {
            error: true,
            helperText: `Too Short, Must be atleast ${
              e[e.length - 1]
            } character(s) long`,
          };
        } else if (e.includes("is not of a type(s) string")) {
          errors[prop as keyof object] = {
            error: true,
            helperText: "Please Enter",
          };
        } else {
          errors[prop as keyof object] = {
            error: true,
          };
        }
      }
    }
  }

  return errors;
}

export default parseFormErrors;
