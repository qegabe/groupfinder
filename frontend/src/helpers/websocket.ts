const host =
  process.env.NODE_ENV === "production"
    ? window.location.host
    : "localhost:5000";

function startWebSocket(
  id: number,
  token: string,
  onMessageCallback: (data: any) => void
): WebSocket {
  const ws = new WebSocket(`ws://${host}/chat/${id}`);

  ws.onopen = () => {
    ws.send(JSON.stringify({ type: "join", token }));
    //console.log("opened ws connection");
  };

  ws.onclose = (e) => {
    console.log("close ws connection:", e.code, e.reason);
  };

  ws.onmessage = (e) => {
    onMessageCallback(e.data);
  };

  return ws;
}

export { startWebSocket };
