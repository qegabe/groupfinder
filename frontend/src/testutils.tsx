import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { RootState, setupStore } from "./store";
import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { PreloadedState } from "@reduxjs/toolkit";

interface CustomRenderOptions {
  currentRoute?: string;
  routePath?: string;
  preloadedState?: PreloadedState<RootState>;
}

function customRender(
  ui: ReactElement,
  cOptions: CustomRenderOptions = {},
  options?: Omit<RenderOptions, "wrapper">
) {
  const store = setupStore(cOptions.preloadedState);
  function AllTheProviders({ children }: { children: React.ReactNode }) {
    return (
      <Provider store={store}>
        <MemoryRouter
          initialEntries={
            cOptions.currentRoute ? [cOptions.currentRoute] : undefined
          }>
          <LocalizationProvider dateAdapter={AdapterDayjs}>
            {cOptions.routePath ? (
              <Routes>
                <Route path={cOptions.routePath} element={children} />
              </Routes>
            ) : (
              children
            )}
          </LocalizationProvider>
        </MemoryRouter>
      </Provider>
    );
  }
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
