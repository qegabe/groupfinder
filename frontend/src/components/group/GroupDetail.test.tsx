import React from "react";
import { render, waitFor, screen, fireEvent, within } from "../../testutils";
import "@testing-library/jest-dom";
import GroupFinderApi from "../../api";
import { WS } from "jest-websocket-mock";
import * as WSHelper from "../../helpers/websocket";
import dayjs from "dayjs";
import GroupDetail from "./GroupDetail";

const mockGetGroup = jest
  .spyOn(GroupFinderApi, "getGroup")
  .mockImplementation(async (id) => {
    return {
      id: 1,
      title: "Test Group",
      description: "test",
      startTime: dayjs("2023-06-23T16:00:00.000Z"),
      endTime: dayjs("2023-06-23T17:00:00.000Z"),
      address: "123 Test Street",
      city: "Test City",
      cityId: 1,
      isPrivate: false,
      maxMembers: 10,
      members: {
        TestUser: {
          avatarUrl: null,
          isOwner: true,
        },
        TestUser3: {
          avatarUrl: null,
          isOwner: false,
        },
      },
      games: [
        { id: 1, title: "Test Game 1", coverUrl: "" },
        { id: -1, title: "Trivia", coverUrl: "" },
      ],
    };
  });

const mockJoinGroup = jest
  .spyOn(GroupFinderApi, "joinGroup")
  .mockImplementation(async () => {});

const mockLeaveGroup = jest
  .spyOn(GroupFinderApi, "leaveGroup")
  .mockImplementation(async () => {});

let WSServer: WS;

const mockStartWebSocket = jest
  .spyOn(WSHelper, "startWebSocket")
  .mockImplementation((path, id, token, onMessageCallback) => {
    const client = new WebSocket("ws://localhost:1234");
    client.onmessage = (e) => {
      onMessageCallback(e.data);
    };
    return client;
  });

beforeEach(() => {
  WSServer = new WS("ws://localhost:1234");
});

afterEach(() => {
  jest.clearAllMocks();
  WS.clean();
});

it("renders", async () => {
  const { container } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  expect(mockGetGroup).toHaveBeenCalledWith(1);
  expect(screen.getByText("Test Group")).toBeInTheDocument();
});

it("matches snapshot", async () => {
  const { container, asFragment } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  expect(asFragment()).toMatchSnapshot();
});

it("shows play trivia if trivia game in group and user is member", async () => {
  const { container } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  expect(screen.getByText("Play Trivia"));
});

it("lets user join group if not member", async () => {
  const { container } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser2", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  const joinButton = screen.getByText("Join Group");
  fireEvent.click(joinButton);
  await waitFor(() => {
    expect(joinButton).not.toBeInTheDocument();
  });
  expect(mockJoinGroup).toHaveBeenCalledWith(1);
});

it("lets user leave group if member", async () => {
  const { container } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser3", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  const leaveButton = screen.getByText("Leave Group");
  fireEvent.click(leaveButton);
  await waitFor(() => {
    expect(leaveButton).not.toBeInTheDocument();
  });
  expect(mockLeaveGroup).toHaveBeenCalledWith(1);
});

it("shows edit button if owner", async () => {
  const { container } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  expect(screen.getByLabelText("Edit"));
});

it("connects to chatroom websocket server", async () => {
  const { container } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  await WSServer.connected;
  expect(mockStartWebSocket).toHaveBeenCalledWith(
    "chat",
    1,
    "",
    expect.any(Function)
  );
});

it("displays messages received from the chatroom", async () => {
  const { container } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  await WSServer.connected;
  const chatButton = screen.getByLabelText("chat");
  fireEvent.click(chatButton);
  WSServer.send(
    JSON.stringify({ type: "system", text: "Test System Message" })
  );
  WSServer.send(
    JSON.stringify({
      type: "chat",
      username: "TestUser",
      text: "Test Chat Message",
      avatarUrl: null,
    })
  );
  await waitFor(() => {
    expect(screen.getByText("Test Chat Message")).toBeInTheDocument();
  });
  expect(screen.getByText("Test System Message")).toBeInTheDocument();
});

it("lets user send messages to the chatroom", async () => {
  const { container } = render(<GroupDetail />, {
    currentRoute: "/groups/1",
    routePath: "/groups/:id",
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  await WSServer.connected;
  const chatButton = screen.getByLabelText("chat");
  fireEvent.click(chatButton);
  const chatInput = screen.getByTestId("chat-input");
  fireEvent.change(chatInput, { target: { value: "Test Sent Message" } });
  fireEvent.submit(chatInput.parentElement as HTMLElement);
  await waitFor(() => {
    expect(WSServer).toReceiveMessage(
      JSON.stringify({
        text: "Test Sent Message",
        username: "TestUser",
        type: "chat",
        avatarUrl: null,
      })
    );
  });
});
