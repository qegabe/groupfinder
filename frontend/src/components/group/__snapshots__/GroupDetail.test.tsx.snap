// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-[object Object] css-18yc32f-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-jfs0iv-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item css-1taorns-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-70qvj9"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3 css-118bm5y-MuiTypography-root"
          >
            Test Group
          </h3>
        </div>
        <div
          class="MuiBox-root css-1jcfcth"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-1e5jr70-MuiGrid-root"
      >
        <hr />
        <div
          class="MuiBox-root css-0"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Location: 123 Test Street, Test City
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Starting at: June 23, 2023 12:00 PM
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Lasts an hour
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-1e5jr70-MuiGrid-root"
      >
        <hr />
        <div
          class="MuiBox-root css-1ebprri"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-p3gdy2-MuiTypography-root"
          >
            test
          </p>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-md-4 css-q299g3-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-1e5jr70-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-1ebprri"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
          >
            <div
              aria-controls="users-content"
              aria-expanded="false"
              class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
              id="users-header"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  Users
                </p>
              </div>
              <div
                class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                >
                  <div
                    aria-labelledby="users-header"
                    class="MuiAccordion-region"
                    id="users-content"
                    role="region"
                  >
                    <div
                      class="MuiAccordionDetails-root css-wmaky3-MuiAccordionDetails-root"
                    >
                      <div
                        class="MuiStack-root css-1p5q5e5-MuiStack-root"
                      >
                        <div
                          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-gr0cqf-MuiPaper-root-MuiCard-root"
                          data-testid="usercard-TestUser"
                        >
                          <div
                            class="MuiCardContent-root css-1u0k5eo-MuiCardContent-root"
                          >
                            <div
                              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-hqywrk-MuiAvatar-root"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                                data-testid="PersonIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                                />
                              </svg>
                            </div>
                            <a
                              class="MuiTypography-root MuiTypography-h6 css-1805zfj-MuiTypography-root"
                              href="/users/TestUser"
                            >
                              TestUser
                            </a>
                          </div>
                        </div>
                        <div
                          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-gr0cqf-MuiPaper-root-MuiCard-root"
                          data-testid="usercard-TestUser3"
                        >
                          <div
                            class="MuiCardContent-root css-1u0k5eo-MuiCardContent-root"
                          >
                            <div
                              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-hqywrk-MuiAvatar-root"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                                data-testid="PersonIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                                />
                              </svg>
                            </div>
                            <a
                              class="MuiTypography-root MuiTypography-h6 css-1805zfj-MuiTypography-root"
                              href="/users/TestUser3"
                            >
                              TestUser3
                            </a>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
          >
            <div
              aria-controls="games-content"
              aria-expanded="false"
              class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
              id="games-header"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  Games
                </p>
              </div>
              <div
                class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                >
                  <div
                    aria-labelledby="games-header"
                    class="MuiAccordion-region"
                    id="games-content"
                    role="region"
                  >
                    <div
                      class="MuiAccordionDetails-root css-wmaky3-MuiAccordionDetails-root"
                    >
                      <div
                        class="MuiStack-root css-76scyl-MuiStack-root"
                      >
                        <div
                          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-gr0cqf-MuiPaper-root-MuiCard-root"
                          data-testid="gamecard-1"
                        >
                          <div
                            class="MuiCardContent-root css-1u0k5eo-MuiCardContent-root"
                          >
                            <div
                              class="MuiAvatar-root MuiAvatar-square css-137moi2-MuiAvatar-root"
                            >
                              <img
                                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                                src="/game-controller.png"
                              />
                            </div>
                            <h6
                              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                            >
                              Test Game 1
                            </h6>
                          </div>
                          <div
                            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
                          />
                        </div>
                        <div
                          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-gr0cqf-MuiPaper-root-MuiCard-root"
                          data-testid="gamecard--1"
                        >
                          <div
                            class="MuiCardContent-root css-1u0k5eo-MuiCardContent-root"
                          >
                            <div
                              class="MuiAvatar-root MuiAvatar-square css-137moi2-MuiAvatar-root"
                            >
                              <img
                                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                                src="/game-controller.png"
                              />
                            </div>
                            <h6
                              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                            >
                              Trivia
                            </h6>
                          </div>
                          <div
                            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
